File: src/test/java/com/koddy/server/global/utils/UniversityInfoTest.java
package com.koddy.server.global.utils;

import com.koddy.server.auth.domain.model.Authenticated;
import com.koddy.server.common.UnitTest;
import com.koddy.server.global.exception.GlobalException;
import com.koddy.server.member.domain.model.Member;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import static com.koddy.server.common.fixture.MentorFixture.MENTOR_1;
import static com.koddy.server.global.exception.GlobalExceptionCode.NOT_PROVIDED_UNIV_DOMAIN;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.junit.jupiter.api.Assertions.assertAll;
import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;

@DisplayName("Member/Mentor -> 도메인 [UniversityInfo] 테스트")
class UniversityInfoTest extends UnitTest {
    private final Member<?> member = MENTOR_1.toDomain().apply(1L);
    private final Authenticated authenticated = new Authenticated(member.getId(), member.getAuthority());

    @Test
    @DisplayName("파악할 수 없는 대학교 도메인은 검증할 수 없다")
    void throwExceptionByNotProvidedUnivDomain() {
        assertAll(
                () -> assertThatThrownBy(() -> UniversityInfo.validateDomain(authenticated, "sjiwon@kyonggi.edu"))
                        .isInstanceOf(GlobalException.class)
                        .hasMessage(NOT_PROVIDED_UNIV_DOMAIN.getMessage()),
                () -> assertThatThrownBy(() -> UniversityInfo.validateDomain(authenticated, "sjiwon@kgu.edu"))
                        .isInstanceOf(GlobalException.class)
                        .hasMessage(NOT_PROVIDED_UNIV_DOMAIN.getMessage()),
                () -> assertThatThrownBy(() -> UniversityInfo.validateDomain(authenticated, "sjiwon@kaya.edu"))
                        .isInstanceOf(GlobalException.class)
                        .hasMessage(NOT_PROVIDED_UNIV_DOMAIN.getMessage()),
                () -> assertThatThrownBy(() -> UniversityInfo.validateDomain(authenticated, "gachon@snu.edu"))
                        .isInstanceOf(GlobalException.class)
                        .hasMessage(NOT_PROVIDED_UNIV_DOMAIN.getMessage())
        );
    }

    @Test
    @DisplayName("대학교 도메인을 검증한다")
    void success() {
        assertAll(
                () -> assertDoesNotThrow(() -> UniversityInfo.validateDomain(authenticated, "sjiwon@kyonggi.ac.kr")),
                () -> assertDoesNotThrow(() -> UniversityInfo.validateDomain(authenticated, "sjiwon@kgu.ac.kr")),
                () -> assertDoesNotThrow(() -> UniversityInfo.validateDomain(authenticated, "sjiwon@kaya.ac.kr")),
                () -> assertDoesNotThrow(() -> UniversityInfo.validateDomain(authenticated, "gachon@snu.ac.kr"))
        );
    }
}

File: src/test/java/com/koddy/server/global/utils/UniversityInfoTest.kt
package com.koddy.server.global.utils

import com.koddy.server.auth.domain.model.Authenticated
import com.koddy.server.common.UnitTestKt
import com.koddy.server.common.fixture.MentorFixture
import com.koddy.server.global.exception.GlobalException
import com.koddy.server.global.exception.GlobalExceptionCode
import com.koddy.server.member.domain.model.Member
import io.kotest.assertions.throwables.shouldNotThrowAny
import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.annotation.DisplayName
import io.kotest.core.spec.style.DescribeSpec
import io.kotest.matchers.throwable.shouldHaveMessage

@UnitTestKt
@DisplayName("Global/Utils -> UniversityInfo 테스트")
internal class UniversityInfoTest : DescribeSpec({
    describe("UniversityInfo's validateDomain") {
        val member: Member<*> = MentorFixture.MENTOR_1.toDomain().apply(1L)
        val authenticated = Authenticated(member.id, member.authority)

        context("시스템에서 관리하지 않는 대학교 도메인이면") {
            val domains = listOf(
                "sjiwon@kyonggi.edu",
                "sjiwon@kgu.edu",
                "sjiwon@kaya.edu",
                "gachon@snu.edu",
            )

            it("NOT_PROVIDED_UNIV_DOMAIN 예외가 발생한다") {
                domains.forEach {
                    shouldThrow<GlobalException> {
                        UniversityInfo.validateDomain(authenticated, it)
                    } shouldHaveMessage GlobalExceptionCode.NOT_PROVIDED_UNIV_DOMAIN.message
                }
            }
        }

        context("시스템에서 관리하고 있는 대학교 도메인이면") {
            val domains = listOf(
                "sjiwon@kyonggi.ac.kr",
                "sjiwon@kgu.ac.kr",
                "sjiwon@kaya.ac.kr",
                "gachon@snu.ac.kr",
            )

            it("검증에 성공한다") {
                domains.forEach {
                    shouldNotThrowAny {
                        UniversityInfo.validateDomain(authenticated, it)
                    }
                }
            }
        }
    }
})
